src/output-json-pop3.c

```diff
@@ -0,0 +1,168 @@
+/* Copyright (C) 2018-2021 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+/**
+ * \file
+ *
+ * \author Alex Savage <alexander.savage@cyber.gc.ca>
+ *
+ * Implement JSON/eve logging app-layer POP3.
+ */
+
+#include "suricata-common.h"
+#include "detect.h"
+#include "pkt-var.h"
+#include "conf.h"
+
+#include "threads.h"
+#include "threadvars.h"
+#include "tm-threads.h"
+
+#include "util-unittest.h"
+#include "util-buffer.h"
+#include "util-debug.h"
+#include "util-byte.h"
+
+#include "output.h"
+#include "output-json.h"
+
+#include "app-layer.h"
+#include "app-layer-parser.h"
+
+#include "app-layer-pop3.h"
+#include "output-json-pop3.h"
+#include "rust.h"
+
+typedef struct LogPOP3FileCtx_ {
+    uint32_t flags;
+    OutputJsonCtx *eve_ctx;
+} LogPOP3FileCtx;
+
+typedef struct LogPOP3LogThread_ {
+    LogPOP3FileCtx *pop3log_ctx;
+    OutputJsonThreadCtx *ctx;
+} LogPOP3LogThread;
+
+static int JsonPOP3Logger(ThreadVars *tv, void *thread_data, const Packet *p, Flow *f, void *state,
+        void *tx, uint64_t tx_id)
+{
+    SCLogNotice("JsonPOP3Logger");
+    LogPOP3LogThread *thread = thread_data;
+
+    JsonBuilder *js =
+            CreateEveHeader(p, LOG_DIR_PACKET, "pop3", NULL, thread->pop3log_ctx->eve_ctx);
```

Comment:
The logger is called when the TX is complete, then LOG_DIR_PACKET logs the ip tuple in the packet direction, so src is the server, dst is the client. For bidirectional transactions like in this parser the src should be the client. To do so, use LOG_DIR_FLOW here.